apps/steam/steam.star

```diff
@@ -0,0 +1,160 @@
+"""
+Applet: Steam
+Summary: Steam Now Playing
+Description: Displays the game that the specified user is currently playing, or the most recent games if currently not in-game.
+Author: Jeremy Tavener
+"""
+
+load("render.star", "render")
+load("http.star", "http")
+load("encoding/base64.star", "base64")
+load("encoding/json.star", "json")
+load("secret.star", "secret")
+load("schema.star", "schema")
+
+API_BASE_URL = "http://api.steampowered.com/"
+API_PLAYER_SUMMARIES = API_BASE_URL + "ISteamUser/GetPlayerSummaries/v0002"
+API_RECENTLY_PLAYED_GAMES = API_BASE_URL + "IPlayerService/GetRecentlyPlayedGames/v0001"
+API_OWNED_GAMES = API_BASE_URL + "IPlayerService/GetOwnedGames/v0001"
+
+def main(config):
+    steam_id = config.get("steam_id", None)
+    api_key = secret.decrypt("""
+        AV6+xWcE+DYxl3gCCIpeRaoTWtnpTqIEqS8zWqnvoStfp/IrLru54NPMHgekh8jwUshSLUFp4Y9GFy7C/
+        zLyBm1DzJODRDfTo3NROgQt7jJI6nzK/8I6MPMv33VZ4/ZjcSH3BkQ0THHB0XcGIv++ntIj3/tcDsvQtr
+        NkoszNXck6JEZaDpo=
+        """) or config.get("dev_api_key")
+
+    if steam_id == None or api_key == None:
+        return display_failure("Steam ID not provided.")
+
+    # Is the user currently playing a game?
+    # Note - this will only return if their profile is set to show this information publically
+    resp = http.get(API_PLAYER_SUMMARIES, params = {"key": api_key, "steamids": steam_id})
+
+    if resp.status_code != 200:
+        return display_failure("Failed to get the current player summary with code {}".format(resp.status_code))
+
+    if len(resp.json()["response"]["players"]) != 1:
+        return display_failure("Failed to find player with SteamID {}".format(steam_id))
+
+    if resp.json()["response"]["players"][0]["communityvisibilitystate"] != 3:
+        return display_failure("Profile is not public, can't get current game")
+
+    player_name = resp.json()["response"]["players"][0]["personaname"]
+
+    if "gameextrainfo" in resp.json()["response"]["players"][0].keys():
+        game_string = "Now Playing: " + resp.json()["response"]["players"][0]["gameextrainfo"]
+        current_game_id = resp.json()["response"]["players"][0]["gameid"]
+
+        # Grab the game Icon - this is groooooosss
+        json_blob = json.encode({"steamid": steam_id, "include_appinfo": True, "appids_filter": [current_game_id]})
+
+        resp = http.get(API_OWNED_GAMES, params = {"key": api_key, "input_json": str(json_blob)})
+
+        if resp.status_code != 200:
+            return display_failure("Failed to get the current game icon with code {}".format(resp.status_code))
+
+        if resp.json()["response"]["game_count"] == 1:
+            game_icon_hash = resp.json()["response"]["games"][0]["img_icon_url"]
+            game_icon_url = "http://media.steampowered.com/steamcommunity/public/images/apps/" + str(current_game_id) + "/" + game_icon_hash + ".jpg"
+            main_icon = http.get(game_icon_url).body()
```

Comment:
Can you please cache this icon? Here's the documentation for the cache module:
https://github.com/tidbyt/pixlet/blob/main/docs/modules.md#pixlet-module-cache

Reply:
I would love to, but I'm slightly at a loss for how to do this with the jpeg it downloads.

I have this for the key: `key = steam_id + "_" + current_game_id` - so it should be unique for the player and the current game.

I'm doing this for pulling the icon and trying to put it in the cache:
`main_icon = http.get(game_icon_url).body()
print("Setting Cache")
cache.set(key, base64.encode(main_icon), ttl_seconds = CACHE_TTL_SECONDS)`

But when I look at the `main_icon`, it looks to be a binary object.

Reply:
`base64.encode(main_icon)` is encoding the binary `main_icon` into a string for the cache.  You'll need to `base64.decode(cache.get(key))` to decode it when you get it out of the cache.

Reply:
Never mind - this was an issue with using `--watch` with `pixlet serve`.  Code works as expected.